airbyte-integrations/connectors/source-snowflake/source_snowflake/streams/table_stream.py

```diff
@@ -231,9 +231,12 @@ def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapp
                                                             for row_type in
                                                             response_json.get('resultSetMetaData', {'rowType': []}).get('rowType', [])])
 
+        CurrentTimeZoneStream.set_off_set(self.url_base, self.config, self.authenticator)
```

Comment:
I proposed this solution because we had multiple post requests to get time zone (one of your previous remarks). This solution is supposed to cache the result. So if the variable is no set, we set it with a post request. Otherwise, no action is taken. Is there an airbyte practice to do thread safe caching in this situation ?

Reply:
I updated the class code to implement lock behavior on the class when setting the value of offset

Reply:
It is as we are moving to have concurrent CDK implemented (as in this project). Can you please use some kind of abstraction where the caching would occur so that the management of this cache is at one place only? I can see something like:

```
class TimeZoneDao(ABC):
  def fetch_timezone() -> <the proper type representing the timezone>:
    raise NotImplemented()

class StreamTimeZoneDao(TimeZoneDao):
  def fetch_timezone() -> <the proper type representing the timezone>:
    <something similar to `next(current_time_zone_stream.read_records(SyncMode.full_refresh))`>
    
class CachingTimeZoneDaoDecorator(TimeZoneDao):
  def fetch_timezone() -> <the proper type representing the timezone>:
    if self._cached_value:
      return self._cached_value
    return self._decorated.fetch_timezone()
```

Reply:
The problem is not only for concurrency but also for testing: now that we rely on a class variable, a test running before another can provide a very different result. We can't rely on class variables for that

Reply:
@maxi297 
I am currently implementing your suggestion, I need some clarification. For the class in your example, CachingTimeZoneDaoDecorator implements fetch_time_zone as an instance method (it is using self), therefore i will have to instantiate CachingTimeZoneDaoDecorator everytime I use it, thus the first call will always send a post request. We will meet again the same issue, of having many requests. 

I am suggesting to set the timezone as instance variable of my streams at the method "streams" in SourceSnowflake. Therefore, there will be no concurrency issues or any caching.

Reply:
Is there a reason you can't instantiate one instance of `CachingTimeZoneDaoDecorator` and use the same instance in all the connectors?

Reply:
I implemented a solution that respects both contraints of being thread safe and not making request.